crates/bifrost/src/providers/replicated_loglet/tasks/repair_tail.rs

```diff
@@ -0,0 +1,274 @@
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::time::Duration;
+
+use restate_core::network::{Networking, TransportConnect};
+use restate_core::{ShutdownError, TaskCenter};
+use restate_types::logs::{KeyFilter, LogletOffset, RecordCache, SequenceNumber};
+use restate_types::net::log_server::{GetDigest, LogServerRequestHeader};
+use restate_types::replicated_loglet::{EffectiveNodeSet, ReplicatedLogletParams};
+use tokio::task::JoinSet;
+use tracing::{trace, warn};
+
+use crate::loglet::util::TailOffsetWatch;
+use crate::providers::replicated_loglet::read_path::ReadStreamTask;
+use crate::providers::replicated_loglet::rpc_routers::LogServersRpc;
+
+use super::digests::Digests;
+
+/// # Overview
+///
+/// The primary reason for why we need such a procedure is because the sequencer node marks the record as
+/// released _before_ it sends out a Release message to log-servers. In other words, we don't require a
+/// quorum of nodes to acknowledge the `Release` message to consider a record committed. This is an important
+/// design optimization to reduce latency of the write path. The sequencer needs 1-RTT (store wave) to commit
+/// a record rather than 2-RTT (store+release).
+///
+/// The tradeoff is the speed at which we can determine the last global committed tail if the sequencer is not
+/// available. There is a number of optimizations that can be added to improve the efficiency of this operation,
+/// those include but are not limited to:
+/// 1. Log-servers persisting the last known_global_tail periodically/async and using this value as known_global_tail on startup.
+/// 2. Sequencer-driven seal. If the sequencer is alive, it can send a special value with the seal message to
+///    indicate what is the ultimate known-global-tail that nodes should repair to instead of relying on the observed max-tail.
+/// 3. Limit `from_offset` to repair from to max(min(local_tails), max(known_global_tails), known_archived, trim_point)
+/// 4. Archiving records to the external object-store instead of re-replication, although, this
+///    doesn't improve the efficiency of the repair operation itself.
+///
+/// The repair process must be reentrant, fault-tolerant, and yields correct immutable state of the
+/// tail even with concurrent runs.
+///
+/// # Details
+///
+/// ## Digest Phase (where are my records?)
+/// Given the input range [max(known_global_tails)..max-local-tail], determine the biggest offset at which we can
+/// find evidence of a write-quorum of a record, including unauthoritative nodes (`StorageState::DataLoss`).
+/// Let's call this `max_known_durable_offset`. We narrow down the repair range to [max_known_durable_offset+1..max-local-tail).
```

Comment:
This is possible because the sequencer waits writing record with lsn `x` until the global tail has reached `x`, right?

Reply:
Yes